toolkit/tools/graphanalytics/graphanalytics.go

```diff
@@ -0,0 +1,295 @@
+
+	"gonum.org/v1/gonum/graph"
+	"gonum.org/v1/gonum/graph/traverse"
+	"gopkg.in/alecthomas/kingpin.v2"
+	"microsoft.com/pkggen/internal/exe"
+	"microsoft.com/pkggen/internal/logger"
+	"microsoft.com/pkggen/internal/pkggraph"
+	"microsoft.com/pkggen/internal/sliceutils"
+)
+
+const (
+	defaultMaxResults = "10"
+)
+
+// mapPair represents a key/value pair in a map[string][]string.
+// What the key and value represent are defined by the functions that use this.
+type mapPair struct {
+	key   string
+	value []string
+}
+
+var (
+	app            = kingpin.New("graphanalytics", "A tool to print analytics of a given dependency graph.")
+	inputGraphFile = exe.InputFlag(app, "Path to the DOT graph file to analyze.")
+	maxResults     = app.Flag("max-results", "The number of results to print per catagory. Set 0 to print unlimited.").Default(defaultMaxResults).Int()
+	logFile        = exe.LogFileFlag(app)
+	logLevel       = exe.LogLevelFlag(app)
+)
+
+func main() {
+	app.Version(exe.ToolkitVersion)
+	kingpin.MustParse(app.Parse(os.Args[1:]))
+
+	logger.InitBestEffort(*logFile, *logLevel)
+
+	err := analyzeGraph(*inputGraphFile, *maxResults)
+	if err != nil {
+		logger.Log.Fatalf("Unable to analyze dependency graph, error: %s", err)
+	}
+}
+
+// analyzeGraph analyzes and prints various attributes of a graph file.
+func analyzeGraph(inputFile string, maxResults int) (err error) {
+	pkgGraph := pkggraph.NewPkgGraph()
+	err = pkggraph.ReadDOTGraphFile(pkgGraph, inputFile)
+	if err != nil {
+		return
+	}
+
+	printDirectlyMostUnresolved(pkgGraph, maxResults)
+	printDirectlyClosestToBeingUnblocked(pkgGraph, maxResults)
+
+	printIndirectlyMostUnresolved(pkgGraph, maxResults)
+	printIndirectlyClosestToBeingUnblocked(pkgGraph, maxResults)
+
+	return
+}
+
+// printIndirectlyMostUnresolved will print the top unresolved packages that are indirectly most blocking.
+func printIndirectlyMostUnresolved(pkgGraph *pkggraph.PkgGraph, maxResults int) {
+	unresolvedPackageDependents := make(map[string][]string)
+
+	for _, node := range pkgGraph.AllNodes() {
+		if node.Type != pkggraph.TypeRun && node.Type != pkggraph.TypeBuild {
+			continue
+		}
+
+		if node.State == pkggraph.StateUnresolved {
+			continue
+		}
+
+		dependentName := nodeDependencyName(node)
+
+		search := traverse.BreadthFirst{}
+		search.Walk(pkgGraph, node, func(n graph.Node, d int) (stopSearch bool) {
+			dependencyNode := n.(*pkggraph.PkgNode)
+
+			if dependencyNode.State == pkggraph.StateUnresolved {
+				unresolvedPkgName := dependencyNode.VersionedPkg.Name
+				insertIfMissing(unresolvedPackageDependents, unresolvedPkgName, dependentName)
+			}
+
+			return
+		})
+	}
+
+	printTitle("[INDIRECT] Most common unresolved dependencies")
+	printMap(unresolvedPackageDependents, "total dependents", maxResults)
+
+	return
+}
+
+// printDirectlyMostUnresolved will print the top unresolved packages that are directly most blocking.
+func printDirectlyMostUnresolved(pkgGraph *pkggraph.PkgGraph, maxResults int) {
```

Comment:
>printDirectlyMostUnresolved [](start = 5, length = 27)

I'm a bit confused by what's happening here vs. `printIndirectlyMostUnresolved` above. The way I understood it, these two should return similar result with the difference that this one here only enumerates each node's direct (unresolved) children, while the other one takes into account the grandchildren as well. From the code it seems that we're actually adding node names through `insertIfMissing` in reverse order. Maybe it's going to become clearer in the morning or after you add some more comments you've mentioned earlier.